Move ResourceError generation from LocalFrameLoaderClient to LoaderStrategy

achristensen07 opened the pull request:
#### db39f001ad27b2d3218a7bbe7fa6e04c985a9850
<pre>
Move ResourceError generation from LocalFrameLoaderClient to LoaderStrategy
<a href="https://bugs.webkit.org/show_bug.cgi?id=285889">https://bugs.webkit.org/show_bug.cgi?id=285889</a>
<a href="https://rdar.apple.com/142845133">rdar://142845133</a>

Reviewed by Pascoe.

WebKit and WebKitLegacy generate slightly different errors.  This is the intent
of a virtual call up the framework stack.  However, ResourceLoader might not
be hooked up to a Frame or a FrameLoaderClient, in which case it would try to
dereference null and crash.  The error generators don&apos;t need the additional
context of a FrameLoaderClient.  Move them to a virtual call on an object that
is never null and never missing, the LoaderStrategy.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
(WebCore::DocumentLoader::interruptedForPolicyChangeError const):
(WebCore::DocumentLoader::startLoadingMainResource):
(WebCore::DocumentLoader::loadMainResource):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::cancelledError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebCore::EmptyFrameLoaderClient::cannotShowURLError const): Deleted.
(WebCore::EmptyFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebCore::EmptyFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebCore::EmptyFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebCore::EmptyFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebCore::EmptyFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebCore::EmptyFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::cancelledError):
(WebCore::FrameLoader::blockedByContentBlockerError):
(WebCore::FrameLoader::blockedError):
(WebCore::FrameLoader::blockedByContentFilterError):
(WebCore::FrameLoader::cancelledError const): Deleted.
(WebCore::FrameLoader::blockedByContentBlockerError const): Deleted.
(WebCore::FrameLoader::blockedError const): Deleted.
(WebCore::FrameLoader::blockedByContentFilterError const): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/LoaderStrategy.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NetscapePlugInStreamLoader.cpp:
(WebCore::NetscapePlugInStreamLoader::didReceiveResponse):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::PolicyChecker::checkNavigationPolicy):
(WebCore::PolicyChecker::cannotShowMIMEType):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::cancelledError):
(WebCore::ResourceLoader::blockedError):
(WebCore::ResourceLoader::blockedByContentBlockerError):
(WebCore::ResourceLoader::cannotShowURLError):
(WebCore::ResourceLoader::httpsUpgradeRedirectLoopError):
* Source/WebCore/loader/appcache/ApplicationCacheHost.cpp:
(WebCore::ApplicationCacheHost::maybeLoadSynchronously):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::cancelledError const):
(WebKit::WebLoaderStrategy::blockedError const):
(WebKit::WebLoaderStrategy::blockedByContentBlockerError const):
(WebKit::WebLoaderStrategy::cannotShowURLError const):
(WebKit::WebLoaderStrategy::interruptedForPolicyChangeError const):
(WebKit::WebLoaderStrategy::blockedByContentFilterError const):
(WebKit::WebLoaderStrategy::cannotShowMIMETypeError const):
(WebKit::WebLoaderStrategy::fileDoesNotExistError const):
(WebKit::WebLoaderStrategy::httpsUpgradeRedirectLoopError const):
(WebKit::WebLoaderStrategy::httpNavigationWithHTTPSOnlyError const):
(WebKit::WebLoaderStrategy::pluginWillHandleLoadError const):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldFallBack const):
(WebKit::WebLocalFrameLoaderClient::cancelledError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::cannotShowURLError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebFrameNetworkingContext.mm:
(WebKit::WebFrameNetworkingContext::blockedError const):
* Source/WebKitLegacy/SourcesCocoa.txt:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.cpp:
(WebResourceLoadScheduler::preconnectTo):
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.mm: Added.
(WebResourceLoadScheduler::pluginWillHandleLoadError const):
(WebResourceLoadScheduler::pluginWillHandleLoadErrorFromResponse):
(WebResourceLoadScheduler::cancelledError const):
(WebResourceLoadScheduler::blockedError const):
(WebResourceLoadScheduler::blockedErrorFromRequest):
(WebResourceLoadScheduler::blockedByContentBlockerError const):
(WebResourceLoadScheduler::cannotShowURLError const):
(WebResourceLoadScheduler::interruptedForPolicyChangeError const):
(WebResourceLoadScheduler::blockedByContentFilterError const):
(WebResourceLoadScheduler::cannotShowMIMETypeError const):
(WebResourceLoadScheduler::fileDoesNotExistError const):
(WebResourceLoadScheduler::httpsUpgradeRedirectLoopError const):
(WebResourceLoadScheduler::httpNavigationWithHTTPSOnlyError const):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::cancelledError const): Deleted.
(WebFrameLoaderClient::blockedError const): Deleted.
(WebFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebFrameLoaderClient::cannotShowURLError const): Deleted.
(WebFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameNetworkingContext.mm:
(WebFrameNetworkingContext::blockedError const):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(-[WebHTMLRepresentation receivedData:withDataSource:]):

Canonical link: <a href="https://commits.webkit.org/289098@main">https://commits.webkit.org/289098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a8f5668596d08096f874fc8883003feecc8bd91a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90466 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36380 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87427 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5162 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13050 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66342 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24154 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88384 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3934 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77514 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46624 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31775 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35448 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74537 "Build is in progress. Recent messages:") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32613 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/91950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9272 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74900 "37 flakes 97 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12914 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74017 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18416 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16849 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13306 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12628 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/18093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12458 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15951 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/loader/NetscapePlugInStreamLoader.cpp:
```diff
@@ -31,7 +31,9 @@
 
 #include "DocumentLoader.h"
 #include "FrameLoader.h"
```
I don't think we need this anymore.

pascoej commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -1073,6 +1073,63 @@ void WebLoaderStrategy::prioritizeResourceLoads(const Vector<WebCore::Subresourc
 
+ResourceError WebLoaderStrategy::cancelledError(const ResourceRequest& request) const
+{
+    return WebKit::cancelledError(request);
```
I think this is in the WebKit namespace and therefore we shouldn't need the prefix.

achristensen07 replied:
We do need the WebKit:: because otherwise it thinks it's recursion.

pascoej commented on Source/WebKit/WebProcess/WebCoreSupport/mac/WebFrameNetworkingContext.mm:
```diff
@@ -66,7 +67,7 @@
 
 ResourceError WebFrameNetworkingContext::blockedError(const ResourceRequest& request) const
 {
-    return protectedFrame()->protectedLoader()->protectedClient()->blockedError(request);
```
"

pascoej reviewed (approved):
Some nits.

achristensen07 committed:
Move ResourceError generation from LocalFrameLoaderClient to LoaderStrategy
https://bugs.webkit.org/show_bug.cgi?id=285889
rdar://142845133

Reviewed by Pascoe.

WebKit and WebKitLegacy generate slightly different errors.  This is the intent
of a virtual call up the framework stack.  However, ResourceLoader might not
be hooked up to a Frame or a FrameLoaderClient, in which case it would try to
dereference null and crash.  The error generators don't need the additional
context of a FrameLoaderClient.  Move them to a virtual call on an object that
is never null and never missing, the LoaderStrategy.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):
(WebCore::DocumentLoader::interruptedForPolicyChangeError const):
(WebCore::DocumentLoader::startLoadingMainResource):
(WebCore::DocumentLoader::loadMainResource):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyFrameLoaderClient::cancelledError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebCore::EmptyFrameLoaderClient::cannotShowURLError const): Deleted.
(WebCore::EmptyFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebCore::EmptyFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebCore::EmptyFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebCore::EmptyFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebCore::EmptyFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebCore::EmptyFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebCore::EmptyFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebCore/loader/EmptyFrameLoaderClient.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::cancelledError):
(WebCore::FrameLoader::blockedByContentBlockerError):
(WebCore::FrameLoader::blockedError):
(WebCore::FrameLoader::blockedByContentFilterError):
(WebCore::FrameLoader::cancelledError const): Deleted.
(WebCore::FrameLoader::blockedByContentBlockerError const): Deleted.
(WebCore::FrameLoader::blockedError const): Deleted.
(WebCore::FrameLoader::blockedByContentFilterError const): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/LoaderStrategy.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/loader/NetscapePlugInStreamLoader.cpp:
(WebCore::NetscapePlugInStreamLoader::didReceiveResponse):
* Source/WebCore/loader/PolicyChecker.cpp:
(WebCore::PolicyChecker::checkNavigationPolicy):
(WebCore::PolicyChecker::cannotShowMIMEType):
* Source/WebCore/loader/ResourceLoader.cpp:
(WebCore::ResourceLoader::cancelledError):
(WebCore::ResourceLoader::blockedError):
(WebCore::ResourceLoader::blockedByContentBlockerError):
(WebCore::ResourceLoader::cannotShowURLError):
(WebCore::ResourceLoader::httpsUpgradeRedirectLoopError):
* Source/WebCore/loader/appcache/ApplicationCacheHost.cpp:
(WebCore::ApplicationCacheHost::maybeLoadSynchronously):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::cancelledError const):
(WebKit::WebLoaderStrategy::blockedError const):
(WebKit::WebLoaderStrategy::blockedByContentBlockerError const):
(WebKit::WebLoaderStrategy::cannotShowURLError const):
(WebKit::WebLoaderStrategy::interruptedForPolicyChangeError const):
(WebKit::WebLoaderStrategy::blockedByContentFilterError const):
(WebKit::WebLoaderStrategy::cannotShowMIMETypeError const):
(WebKit::WebLoaderStrategy::fileDoesNotExistError const):
(WebKit::WebLoaderStrategy::httpsUpgradeRedirectLoopError const):
(WebKit::WebLoaderStrategy::httpNavigationWithHTTPSOnlyError const):
(WebKit::WebLoaderStrategy::pluginWillHandleLoadError const):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::shouldFallBack const):
(WebKit::WebLocalFrameLoaderClient::cancelledError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::cannotShowURLError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebKit::WebLocalFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/mac/WebFrameNetworkingContext.mm:
(WebKit::WebFrameNetworkingContext::blockedError const):
* Source/WebKitLegacy/SourcesCocoa.txt:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.cpp:
(WebResourceLoadScheduler::preconnectTo):
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.h:
* Source/WebKitLegacy/WebCoreSupport/WebResourceLoadScheduler.mm: Added.
(WebResourceLoadScheduler::pluginWillHandleLoadError const):
(WebResourceLoadScheduler::pluginWillHandleLoadErrorFromResponse):
(WebResourceLoadScheduler::cancelledError const):
(WebResourceLoadScheduler::blockedError const):
(WebResourceLoadScheduler::blockedErrorFromRequest):
(WebResourceLoadScheduler::blockedByContentBlockerError const):
(WebResourceLoadScheduler::cannotShowURLError const):
(WebResourceLoadScheduler::interruptedForPolicyChangeError const):
(WebResourceLoadScheduler::blockedByContentFilterError const):
(WebResourceLoadScheduler::cannotShowMIMETypeError const):
(WebResourceLoadScheduler::fileDoesNotExistError const):
(WebResourceLoadScheduler::httpsUpgradeRedirectLoopError const):
(WebResourceLoadScheduler::httpNavigationWithHTTPSOnlyError const):
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameLoaderClient.mm:
(WebFrameLoaderClient::cancelledError const): Deleted.
(WebFrameLoaderClient::blockedError const): Deleted.
(WebFrameLoaderClient::blockedByContentBlockerError const): Deleted.
(WebFrameLoaderClient::cannotShowURLError const): Deleted.
(WebFrameLoaderClient::interruptedForPolicyChangeError const): Deleted.
(WebFrameLoaderClient::blockedByContentFilterError const): Deleted.
(WebFrameLoaderClient::cannotShowMIMETypeError const): Deleted.
(WebFrameLoaderClient::fileDoesNotExistError const): Deleted.
(WebFrameLoaderClient::httpsUpgradeRedirectLoopError const): Deleted.
(WebFrameLoaderClient::httpNavigationWithHTTPSOnlyError const): Deleted.
(WebFrameLoaderClient::pluginWillHandleLoadError const): Deleted.
* Source/WebKitLegacy/mac/WebCoreSupport/WebFrameNetworkingContext.mm:
(WebFrameNetworkingContext::blockedError const):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(-[WebHTMLRepresentation receivedData:withDataSource:]):

Canonical link: https://commits.webkit.org/289098@main
